### PR TITLE
New remote write format

### PR DESCRIPTION
**Pull request by cstyan**
@npazosmendez @pracucci

This PR introduces a new "minimized" string interning remote write format based on experiments by Marco (@pracucci). 

The TL;DR is that the new format has a single string containing all the actual symbols, similar to the actual TSDB symbol table, and references are now the offset and length of the substring to read from the symbols string.

The initial benchmark results are promising;
the current RW format
```
BenchmarkBuildWriteRequest/2_instances-32         	    1000	     17427 ns/op	      1933 compressedSize/op	      80 B/op       1 allocs/op
BenchmarkBuildWriteRequest/10_instances-32        	    1000	     85246 ns/op	      8156 compressedSize/op	      96 B/op       1 allocs/op
BenchmarkBuildWriteRequest/1k_instances-32        	    1000	    873437 ns/op	     78215 compressedSize/op	      84 B/op       1 allocs/op
PASS
ok  	github.com/prometheus/prometheus/storage/remote	1.216s
```

the previously proposed "reduced" format
```
BenchmarkBuildReducedWriteRequest/2_instances-32         	    1000	     24873 ns/op	      1804 compressedSize/op	      64 B/op	       1 allocs/op
BenchmarkBuildReducedWriteRequest/10_instances-32        	    1000	    105308 ns/op	      6094 compressedSize/op	      80 B/op	       1 allocs/op
BenchmarkBuildReducedWriteRequest/1k_instances-32        	    1000	    963141 ns/op	     54953 compressedSize/op	      68 B/op	       1 allocs/op
```

and finally the new "minimized" format
```
BenchmarkBuildMinimizedWriteRequest/66-instances-32         	    1000	         9.790 ns/op	      1567 compressedSize/op	      10 B/op	       0 allocs/op
BenchmarkBuildMinimizedWriteRequest/330-instances-32        	    1000	        46.02 ns/op	      5659 compressedSize/op	      50 B/op	       0 allocs/op
BenchmarkBuildMinimizedWriteRequest/3300-instances-32       	    1000	       470.2 ns/op	     53153 compressedSize/op	     508 B/op	       3 allocs/op
```

However, what is a bit concerning are the results of running all 3 versions via Nico and Alex's e2e scripts when we look at the results for the new minimized format. The network bytes and receiver resources are all good. We're sending even less bytes over the wire and the receiver CPU is less than half what it is for the other two formats. What I want to dig into more is the CPU usage of if the minimized format sender is ~10% higher than the current format and the reduced format.

e2e bench screenshots
2023-11-06-175114_2522x842_scrot
2023-11-06-175124_2517x803_scrot

Here's heap and cpu profiles, the only thing I can think might be the cause is the map access or somehow missing the passing + reuse of a buffer somewhere.
[heap.tar.gz](https://github.com/npazosmendez/prometheus/files/13275173/heap.tar.gz)
[cpu.tar.gz](https://github.com/npazosmendez/prometheus/files/13275176/cpu.tar.gz)

### CODE REVIEW

**Comment by npazosmendez** [comment]
Great work, incredible gains on the receiver end. I think it makes sense to try to optimize the sender's CPU on this version and try to push it through.

On the senders CPU: I wasn't able to get benchmark results similar to yours. Probably partially because I had to tweak the code to make it compile with the stringlabels tag, but I also think we are not doing the same measurements. Anyway, I found some minor things:
- `strings.Builder.Reset()` discards the underlying buffer and does not reuse it. We can try what Bryan did [here](https://github.com/prometheus/prometheus/blob/b80617a248a557f6f9e7e708dafb0381ccd414a8/model/labels/labels_stringlabels.go#L404): using a raw byte array and doing an unsafe cast to a string. We need to be cautios about the string encoding, specially because we use the string's length for indexing. In any case, our case is more trivial than the one I linked, so we should be able to figure it out.
- Map lookups are a considerable part of the CPU. I noticed one of them could be avoided [here](https://github.com/prometheus/prometheus/blob/6a6fc571320e8d37a80665e8e9ae653d27db83ea/storage/remote/queue_manager.go#L1892).

With those two optimizations I could get the CPU down a bit: https://github.com/npazosmendez/prometheus/commit/cb3d322bde7128d252f155e2c93f926fe4db4602  (`sender-v11-min-optimized`)

image


More optimizations like this one can probably be done. I don't think there's something about this approach that is considerably more expensive than the other one. They both use a hash map to de-duplicate strings.

I can continue to look into optimizing this. But let's sync offline about next steps.

**Comment by npazosmendez** [comment]
More investigations: looking at the profiles I see v11-minimized's CPU spends much more doing GC than v1. I see a lot of allocations

**Comment by npazosmendez** [comment]
More investigations: looking at the profiles I see v11-minimized's CPU spends much more doing GC than v1. I see a lot of allocations in the marshaling of the minimized request:


image

Most of it is coming from `(*MinimizedTimeSeries).MarshalToSizedBuffer`, specifically this line:

https://github.com/npazosmendez/prometheus/blob/cb3d322bde7128d252f155e2c93f926fe4db4602/prompb/types.pb.go#L2408

<details>
<summary>Image</summary>


image


</details>

I don't know why it needs to create that extra buffer each time and is not reused accross marshal calls. Without digging into it much more, I wonder if  replacing the `repeated uint32 label_symbols'` with a `repeated NameValue label_symbols` structs removes this allocations. See for example what happens with the Samples: https://github.com/npazosmendez/prometheus/blob/cb3d322bde7128d252f155e2c93f926fe4db4602/prompb/types.pb.go#L2393-L2406

EDIT: ah, I see there's `MinimizedLabelSymbol` defined but unused. We could try that.

**Comment by npazosmendez** [comment]
On a separate note: 20 bits to encode the index to the string means it can only be as high as 1048576. Assuming an average of length 8 per string that's ~100k unique possible labels per sample. It sounds reasonable, but just stating it to make it explicit. Are there other limits in prometheus that are already bounding the number of different labels in a sample?

**Comment by npazosmendez** [comment]
And on a note related to the one above: for a fair comparison of this proto vs the one "with references" (original interning), we should limit the number of labels in the same way. Currently

**Comment by npazosmendez** [comment]
Followup on https://github.com/npazosmendez/prometheus/pull/2#issuecomment-1799347846 : I changed the `repeated uint32 label_symbols` (which is encoded as a varint) into a `repeated fixed32 label_symbols` (which is encoded in 32 bits exactly): https://github.com/npazosmendez/prometheus/commit/ca6faccc5e15e8f8b13dfbdf2228cb4aefa0a3f4 . I think the varint encoding was what needed the extra buffer during marshaling. This change tricked the proto generator and it does not allocate that extra memory anymore. It also makes serialization and deserialization faster.

Regarding the payload-size, I don't expect it will be very impacftul; we are _always_ using the first 13 bits of the uint32 anyway, so the varint encoding is bound to that already. Also, regarding network usage, compression will do it's part later.


Anyway, results: no CPU penalty, and sometimes it's even lower, also on the receiver:

image




With this results I think we can discard the current interning approach and go for this one, it's better in all aspects (expect maybe complexity, and maybe the unsafe cast). I will do some final benchmarks to confirm.

**Comment by cstyan** [comment]
here's a run of the new format after Nico's fix over > 1h
2023-11-07-183205_2513x684_scrot
2023-11-07-183212_2513x916_scrot


We have a small negative hit to memory usage on the sender, but other than that cpu usage is more or less the same. The main benefits are the massive decrease to the wire bytes as well as the resource usage on the receiver.

cc @pracucci

**Comment by npazosmendez** [comment]
As discussed internally, a single `fixed32` for both the index and length might be too little and could make the protocol too restrictive. 

I tried encoding them into two different `fixed32` instead of packing them into one, so the both get the full 32 bits. As expected, this increased payload sizes (from previously a ~39% decrease, to a ~25% decrease). Then I decided to insist with varints, I tries with two `uint32`; as expected, payload sizes went down again (~39% decrease) but sender's CPU went up that 10% from before.  So I decided to fight gogoproto and manually edit the memory-inefficient marshaling code, doing the encoding inplace: https://github.com/npazosmendez/prometheus/commit/6c03bfbaae5cb8f1afeda18f8320f6cec510d681


That gave great results: sender's CPU is back to -+1%, and payload sizes down 39%. And we have 32 bits for the length and other 32 bits for the offset. Should be more than enough.

image


We'll need to decide what to do with this hack, editing the generated code is not ideal. But this just proves the performance is possible.

**Comment by cstyan** [comment]
@npazosmendez we should rebase the branch you and alex started `alexnico-remote-write-1-1` to have the e2e benches that have been merged into the main feature branch. We can merge this and then clean up that branch in the near future to only include the one final new proto format we decide on.

I added extended the existing tests to cover the new minimized format as well.

**Comment by npazosmendez** [comment]
Done. I've rebased `alexnico-remote-write-1-1` on top of `remote-write-1.1`, and then this branch on top of `alexnico-remote-write-1-1`. I then merged it locally because I can't push to your branch: https://github.com/npazosmendez/prometheus/commits/alexnico-remote-write-1-1 

I'm closing this PR. We'll continue to cleanup and optimize on the other branch.
